src/app/CommandSender.h

```diff
@@ -181,29 +181,98 @@ class CommandSender final : public Messaging::ExchangeDelegate
+    // to make necessary changes for new functionality in a single location.
+    struct AddRequestDataParameters
     {
         // gcc bug requires us to have the constructor below
         // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
-        AdditionalCommandParameters() {}
+        AddRequestDataParameters() {}
 
-        AdditionalCommandParameters & SetStartOrEndDataStruct(bool aStartOrEndDataStruct)
+        AddRequestDataParameters(const Optional<uint16_t> & aTimedInvokeTimeoutMs) : timedInvokeTimeoutMs(aTimedInvokeTimeoutMs) {}
+
+        // When a value is provided for timedInvokeTimeoutMs, this invoke becomes a timed
+        // invoke. CommandSender will use the minimum of all provided timeouts for execution.
+        const Optional<uint16_t> timedInvokeTimeoutMs;
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response.
+        Optional<uint16_t> commandRef;
+    };
+
+    // PrepareCommand is a public SDK API, so we must maintain source compatibility.
+    // Using this struct for API parameters instead of individual parameters allows us
+    // to make necessary changes for new functionality in a single location.
+    struct PrepareCommandParameters
+    {
+        // gcc bug requires us to have the constructor below
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
+        PrepareCommandParameters() {}
+
+        PrepareCommandParameters(const AddRequestDataParameters & aAddRequestDataParam) :
+            commandRef(aAddRequestDataParam.commandRef)
+        {}
+
+        PrepareCommandParameters & SetStartDataStruct(bool aStartDataStruct)
         {
-            startOrEndDataStruct = aStartOrEndDataStruct;
+            startDataStruct = aStartDataStruct;
             return *this;
         }
 
-        // From the perspective of `PrepareCommand`, this is an out parameter. This value will be
-        // set by `PrepareCommand` and is expected to be unchanged by caller until it is provided
-        // to `FinishCommand`.
+        PrepareCommandParameters & SetCommandRef(uint16_t aCommandRef)
+        {
+            commandRef.SetValue(aCommandRef);
+            return *this;
+        }
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response. We validate the reference
+        // early in PrepareCommand, even though it's not used until FinishCommand. This proactive
+        // validation helps prevent unnecessary writing an InvokeRequest into the packet that later
+        // needs to be undone.
+        //
+        // Currently, provided commandRefs for the first request must start at 0 and increment by one
+        // for each subsequent request. This requirement can be relaxed in the future if a compelling
+        // need arises.
```

Comment:
This requirement should be relaxed, period.  There's no reason to make people provide us with information if it has to be information we already know...

Followup issue for this, please, and reference it here.

Reply:
Added TODO comment here as well as an item in https://github.com/project-chip/connectedhomeip/issues/30453 for tracking purpose.

I will try to have the tracker completed in time for SVE code freeze. But as we discussed this might be a little delayed by a couple weeks depending on priority of other efforts that need to be done before SVE code freeze